File: elixir/recursive.exs
defmodule Solver do
  def solve([curr | map], accum, title_counter, lesson_counter) do
    curr = Map.put(curr, "position", title_counter)
    lesson_counter = if curr["reset_lesson_position"], do: 0, else: lesson_counter

    curr =
      Map.put(
        curr,
        "lessons",
        Enum.map(Enum.with_index(curr["lessons"], 1), fn {lesson, index} ->
          Map.put(lesson, "position", lesson_counter + index)
        end)
      )

    solve(map, [curr | accum], title_counter + 1, lesson_counter + Enum.count(curr["lessons"]))
  end

  def solve([], accum, _title_counter, _lesson_counter) do
    Enum.reverse(accum)
  end
end

input = [
  %{
    "title" => "Getting started",
    "reset_lesson_position" => false,
    "lessons" => [
      %{"name" => "Welcome"},
      %{"name" => "Installation"}
    ]
  },
  %{
    "title" => "Basic operator",
    "reset_lesson_position" => false,
    "lessons" => [
      %{"name" => "Addition / Subtraction"},
      %{"name" => "Multiplication / Division"}
    ]
  },
  %{
    "title" => "Advanced topics",
    "reset_lesson_position" => true,
    "lessons" => [
      %{"name" => "Mutability"},
      %{"name" => "Immutability"}
    ]
  }
]

test = [
  %{
    "title" => "Getting started",
    "reset_lesson_position" => false,
    "position" => 1,
    "lessons" => [
      %{"name" => "Welcome", "position" => 1},
      %{"name" => "Installation", "position" => 2}
    ]
  },
  %{
    "title" => "Basic operator",
    "reset_lesson_position" => false,
    "position" => 2,
    "lessons" => [
      %{"name" => "Addition / Subtraction", "position" => 3},
      %{"name" => "Multiplication / Division", "position" => 4}
    ]
  },
  %{
    "title" => "Advanced topics",
    "reset_lesson_position" => true,
    "position" => 3,
    "lessons" => [
      %{"name" => "Mutability", "position" => 1},
      %{"name" => "Immutability", "position" => 2}
    ]
  }
]

output = Solver.solve(input, [], 1, 0)
IO.inspect(output)
true = output == test

File: ocaml/recursive.ml
module Lesson = struct
  type t = { name : string; position : int option }
end

module Chapter = struct
  type t = {
    title : string;
    reset_lesson_position : bool;
    lessons : Lesson.t list;
    position : int option;
  }
end

let input : Chapter.t list =
  [
    {
      title = "Getting started";
      reset_lesson_position = false;
      lessons =
        [
          { name = "Welcome"; position = None };
          { name = "Installation"; position = None };
        ];
      position = None;
    };
    {
      title = "Basic operator";
      reset_lesson_position = false;
      lessons =
        [
          { name = "Addition / Subtraction"; position = None };
          { name = "Multiplication / Division"; position = None };
        ];
      position = None;
    };
    {
      title = "Advanced topics";
      reset_lesson_position = true;
      lessons =
        [
          { name = "Mutability"; position = None };
          { name = "Immutability"; position = None };
        ];
      position = None;
    };
  ]

(* These functions are tail-call optimized so super efficient *)
let rec process_lessons ?(i = 1) ?(lessons = []) = function
  | [] -> lessons
  | lesson :: rest_lessons ->
      process_lessons ~i:(i + 1)
        ~lessons:({ lesson with Lesson.position = Some i } :: lessons)
        rest_lessons

and process_chapters ?(i = 1) ?(lesson_i = 1) ?(chapters = []) = function
  | [] -> List.rev chapters
  | (chapter : Chapter.t) :: rest_chapters ->
      let lessons = process_lessons ~i:lesson_i chapter.lessons in
      let lesson_i : int =
        match (chapter, lessons) with
        | { reset_lesson_position = true }, _ -> 1
        | _, { position = Some pos } :: _ -> pos
        (* we need to take care of this case, because we reuse the type, even
           though the position is never None *)
        | _, { position = None } :: _ | _, [] -> lesson_i
      in
      process_chapters
        ~chapters:
          ({ chapter with position = Some i; lessons = List.rev lessons }
           :: chapters)
        ~i:(i + 1) ~lesson_i rest_chapters

(* the end result *)
let chapters = process_chapters input
